File: Assignment-3/CPP/AEHelper.cpp
//===- AEHelper.cpp -- Abstract Interpretation harness --//
//
//                     SVF: Static Value-Flow Analysis
//
// Copyright (C) <2013-2022>  <Yulei Sui>
//

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.

// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//
//===----------------------------------------------------------------------===//
/*
 * Harness for Assignment-3 abstract interpretation.
 *
 * Owns the harness-side `AbstractExecution::*` methods that students don't
 * design:
 *   - Interprocedural WTO construction   (initWTO)
 *   - Stub / checkpoint sub-dispatch     (handleStubFunctions /
 *                                         handleCheckpointStubs) — invoked
 *                                         from the student's handleCallSite
 *                                         in Assignment_3.cpp.
 *   - External-API whitelist             (isExternalCallForAssignment)
 *   - Abstract-state helpers             (getAbsValue / updateAbsValue /
 *                                         loadValue / storeValue / GEP* /
 *                                         getAEState / postAbsTrace)
 *   - Validator                          (ensureAllAssertsValidated)
 *
 * Pure bug-reporting concerns (AEReporter class + JSON / coverage summary)
 * live in AEReporter.cpp.  The analysis driver (runOnModule / analyse /
 * handleCallSite / reportBufOverflow / reportNullDeref) and the six student
 * tasks live in Assignment_3.cpp.
 */

#include "Assignment_3.h"
#include "AEState.h"
#include "WPA/Andersen.h"
#include <sstream>

using namespace SVF;

/// Whitelist of external-call names the assignment expects students to model
/// in `updateStateOnExtCall`.  Covers:
///   - Assignment-specific stubs:        `mem_insert`, `str_insert`
///   - Memory family:                    `memcpy`, `memmove`, `memset`
///   - String family:                    `strcpy`, `strncpy`, `strcat`,
///                                       `strncat`, `strlen`, `wcslen`
///   - Ground-truth checkpoint stubs:    `SAFE_/UNSAFE_BUFACCESS`,
///                                       `SAFE_/UNSAFE_PTRDEREF`
///
/// The library APIs are matched by substring because Clang emits the memory
/// family as LLVM intrinsics (e.g. `llvm.memcpy.p0.p0.i64`) and the substring
/// is preserved in the mangled name.
bool AbstractExecution::isExternalCallForAssignment(const SVF::FunObjVar* func) {
	const std::string& name = func->getName();
	static const Set<std::string> exactStubs = {
	    "mem_insert", "str_insert",
	    "UNSAFE_BUFACCESS", "SAFE_BUFACCESS",
	    "UNSAFE_PTRDEREF", "SAFE_PTRDEREF"};
	if (exactStubs.count(name))
		return true;
	static const std::vector<std::string> apiSubstrings = {
	    "memcpy", "memmove", "memset",
	    "strcpy", "strncpy", "strcat", "strncat",
	    "strlen", "wcslen"};
	for (const auto& key : apiSubstrings) {
		if (name.find(key) != std::string::npos)
			return true;
	}
	return false;
}

// ---------------------------------------------------------------------------
// WTO construction.  Each (mutually) recursive function's entry node becomes
// a WTO cycle head because intra-SCC call edges are turned into back-edges.
// The same widening/narrowing machinery used for loops then drives recursion
// to a fixpoint via handleICFGCycle; recursive callsites are filtered out in
// handleCallSite via `inSameCallGraphSCC`.
// ---------------------------------------------------------------------------

void AbstractExecution::initWTO() {
	ander = AndersenWaveDiff::createAndersenWaveDiff(svfir);
	Andersen::CallGraphSCC* callGraphScc = ander->getCallGraphSCC();
	callGraphScc->find();
	auto callGraph = ander->getCallGraph();

	for (auto it = callGraph->begin(); it != callGraph->end(); ++it) {
		const FunObjVar* fun = it->second->getFunction();
		if (fun->isDeclaration())
			continue;

		NodeID repNodeId = callGraphScc->repNode(it->second->getId());
		const NodeBS& cgSCCNodes = callGraphScc->subNodes(repNodeId);

		bool isEntry = it->second->getInEdges().empty();
		for (auto inEdge : it->second->getInEdges())
			if (!cgSCCNodes.test(inEdge->getSrcID()))
				isEntry = true;
		if (!isEntry)
			continue;

		Set<const FunObjVar*> funcScc;
		for (const auto& node : cgSCCNodes)
			funcScc.insert(callGraph->getGNode(node)->getFunction());

		auto* wto = new ICFGWTO(icfg->getFunEntryICFGNode(fun), funcScc);
		wto->init();
		funcToWTO[fun] = wto;
	}
}

/// Verify that every ground-truth stub call site was reached by the student's
/// analysis (added to `assert_points` via handleCallSite -> handleStubFunctions
/// / handleCheckpointStubs).  A missed stub site means the student's
/// control-flow logic skipped a place the grader cares about.
///
/// Recognised stubs:
///   - svf_assert / svf_assert_eq         : abstract-state assertion checks
///   - UNSAFE_PTRDEREF / SAFE_PTRDEREF    : null-deref ground truth
///   - UNSAFE_BUFACCESS / SAFE_BUFACCESS  : buffer-access ground truth
///
/// Additionally requires exact report counts for each checkpoint bug kind.
void AbstractExecution::ensureAllAssertsValidated() {
	static const Set<std::string> kAssertStubs = {"svf_assert", "svf_assert_eq"};
	static const Set<std::string> kCheckpointStubs = {
	    "UNSAFE_PTRDEREF", "SAFE_PTRDEREF",
	    "UNSAFE_BUFACCESS", "SAFE_BUFACCESS"};
	for (auto it = svfir->getICFG()->begin(); it != svfir->getICFG()->end(); ++it) {
		const ICFGNode* node = it->second;
		const CallICFGNode* call = SVFUtil::dyn_cast<CallICFGNode>(node);
		if (!call)
			continue;
		const FunObjVar* fun = call->getCalledFunction();
		if (!fun)
			continue;
		const std::string& name = fun->getName();
		const bool isAssertStub = kAssertStubs.count(name) > 0;
		const bool isCheckpointStub = kCheckpointStubs.count(name) > 0;
		if (!isAssertStub && !isCheckpointStub)
			continue;
		if (!bugReporter.isAssertionPoint(call)) {
			std::stringstream ss;
			ss << "The stub function callsite (" << name
			   << ") was not reached by the student's control flow: "
			   << call->toString();
			std::cerr << ss.str() << std::endl;
			assert(false);
		}
	}

	static const std::vector<std::string> kReportKinds = {
	    "buffer-overflow", "nullptr-deref"};
	for (const std::string& kind : kReportKinds) {
		const u32_t expected = bugReporter.getExpectedReportCount(kind);
		const u32_t actual = bugReporter.getReportCount(kind);
		if (actual != expected) {
			std::cerr << "Assignment 3 report-count mismatch for " << kind
			          << ": expected " << expected << ", got " << actual
			          << std::endl;
			assert(false && "Assignment 3 report count did not match checkpoints");
		}
	}
}

/// Record SAFE/UNSAFE checkpoint expectations. The checkpoint does not inspect
/// abstract state or emit a bug; only the student's checker may report one.
void AbstractExecution::handleCheckpointStubs(const CallICFGNode* callNode) {
	bugReporter.noteAssertionPoint(callNode);
	const std::string fun_name = callNode->getCalledFunction()->getName();
	if (fun_name == "UNSAFE_BUFACCESS")
		bugReporter.noteExpectedReport("buffer-overflow");
	else if (fun_name == "UNSAFE_PTRDEREF")
		bugReporter.noteExpectedReport("nullptr-deref");
}

/// Handle the abstract-state assertion stubs.  `svf_assert(expr)` requires the
/// expression to hold true; `svf_assert_eq(a, b)` requires the two intervals
/// to be equal.  Both record the call site in `assert_points` so
/// `ensureAllAssertsValidated` can verify coverage.
void AbstractExecution::handleStubFunctions(const SVF::CallICFGNode* callNode) {
	if (callNode->getCalledFunction()->getName() == "svf_assert") {
		bugReporter.noteAssertionPoint(callNode);
		u32_t arg0 = callNode->getArgument(0)->getId();
		AEState& as = getAEState(callNode);

		if (as[arg0].getInterval().is_infinite()) {
			SVFUtil::errs() << "svf_assert Fail. " << callNode->toString() << "\n";
			assert(false);
		}
		else {
			if (as[arg0].getInterval().equals(IntervalValue(1, 1))) {
				std::stringstream ss;
				ss << "The assertion (" << callNode->toString() << ")"
				   << " is successfully verified!!\n";
				SVFUtil::outs() << ss.str() << std::endl;
			}
			else {
				std::stringstream ss;
				ss << "The assertion (" << callNode->toString() << ")"
				   << " is unsatisfiable!!\n";
				SVFUtil::outs() << ss.str() << std::endl;
				assert(false);
			}
		}
		return;
	}
	else if (callNode->getCalledFunction()->getName() == "svf_assert_eq")  {
		bugReporter.noteAssertionPoint(callNode);
		u32_t arg0 = callNode->getArgument(0)->getId();
		u32_t arg1 = callNode->getArgument(1)->getId();
		AEState& as = getAEState(callNode);
		if (as[arg0].getInterval().equals(as[arg1].getInterval())) {
			SVFUtil::errs() << SVFUtil::sucMsg("The assertion is successfully verified!!\n");
		}
		else {
			SVFUtil::errs() << "svf_assert_eq Fail. " << callNode->toString() << "\n";
			assert(false);
		}
		return;
	}
}

// ===========================================================================
// Abstract-state helpers owned by Assignment-3.
// ===========================================================================
namespace SVF {

const AbstractValue& AbstractExecution::getAbsValue(const ValVar* var, const ICFGNode* node) {
	return getAEState(node).getAbsValue(var);
}
const AbstractValue& AbstractExecution::getAbsValue(const ObjVar* var, const ICFGNode* node) {
	return getAEState(node).getAbsValue(var);
}
const AbstractValue& AbstractExecution::getAbsValue(const SVFVar* var, const ICFGNode* node) {
	return getAEState(node).getAbsValue(var);
}
void AbstractExecution::updateAbsValue(const ValVar* var, const AbstractValue& val, const ICFGNode* node) {
	getAEState(node).updateAbsValue(var, val);
}
void AbstractExecution::updateAbsValue(const ObjVar* var, const AbstractValue& val, const ICFGNode* node) {
	getAEState(node).updateAbsValue(var, val);
}
void AbstractExecution::updateAbsValue(const SVFVar* var, const AbstractValue& val, const ICFGNode* node) {
	getAEState(node).updateAbsValue(var, val);
}
AbstractValue AbstractExecution::loadValue(const ValVar* pointer, const ICFGNode* node) {
	return getAEState(node).loadValue(pointer);
}
void AbstractExecution::storeValue(const ValVar* pointer, const AbstractValue& val, const ICFGNode* node) {
	getAEState(node).storeValue(pointer, val);
}
AddressValue AbstractExecution::getGepObjAddrs(const ValVar* pointer, IntervalValue offset) {
	return getAEState(pointer->getICFGNode()).getGepObjAddrs(pointer, offset);
}
IntervalValue AbstractExecution::getGepElementIndex(const GepStmt* gep) {
	return getAEState(gep->getICFGNode()).getGepElementIndex(gep);
}
IntervalValue AbstractExecution::getGepByteOffset(const GepStmt* gep) {
	return getAEState(gep->getICFGNode()).getGepByteOffset(gep);
}
u32_t AbstractExecution::getAllocaInstByteSize(const AddrStmt* addr) {
	return getAEState(addr->getICFGNode()).getAllocaInstByteSize(addr);
}

/// CallPE is phi-like: the formal parameter joins the caller-side value from
/// every call site represented by the statement.
void AbstractExecution::updateStateOnCall(const CallPE* callPE) {
	AEState& state = getAEState(callPE->getICFGNode());
	AbstractValue joined;
	for (u32_t index = 0; index < callPE->getOpVarNum(); ++index) {
		const ICFGNode* callNode = callPE->getOpCallICFGNode(index);
		if (postAbsTrace.count(callNode))
			joined.join_with(postAbsTrace[callNode][callPE->getOpVarID(index)]);
	}
	state[callPE->getResID()] = joined;
}

// Assignment-3-owned post-trace accessors.
AEState& AbstractExecution::getAEState(const ICFGNode* node) {
	return postAbsTrace[node];
}

} // namespace SVF

File: Assignment-3/Python/AEHelper.py
"""Harness for Assignment-3 abstract interpretation.

Owns the AbstractExecution class's harness-side methods — the analysis driver,
interprocedural WTO construction (initWto), stub / checkpoint sub-dispatchers
(handleStubFunction, handleCheckpointStubs) invoked from the student's
handleCallSite override in Assignment_3.py, the external-API whitelist
(isExternalCallForAssignment), access to the Assignment-3-owned
abstract-state trace, and the validator
(ensureAllAssertsValidated).

The AEReporter class (pure bug reporting + JSON / coverage summary plus
the GEP / strlen / memcpy helpers used by the bug checkers) lives in
AEReporter.py. The six assignment features live in Assignment_3.py.
"""

from abc import ABC, abstractmethod
import faulthandler

import pysvf
from pysvf import ICFG, AbstractValue, ICFGNode, IntervalValue
from pysvf.enums import Predicate

from AEReporter import AEReporter
from AEState import AEState, unwrap_state

faulthandler.enable()


class WTOCycleDepth:
    def __init__(self):
        self._heads: list[ICFGNode] = []

    def add(self, head: ICFGNode):
        self._heads.append(head)

    def __iter__(self):
        return iter(self._heads)

    def __str__(self):
        return f"{self._heads}"

    def __repr__(self):
        return f"{self._heads}"

    def compare(self, other):
        if self == other:
            return 0
        this_it = iter(self)
        other_it = iter(other)
        while this_it:
            if not other_it:
                return 1
            elif this_it == other_it:
                this_it = next(this_it)
                other_it = next(other_it)
            else:
                return 2
        if not other_it:
            return 0
        else:
            return -1

    def __lt__(self, other):
        return self.compare(other) == -1

    def __le__(self, other):
        return self.compare(other) <= 0

    def __eq__(self, other):
        return self.compare(other) == 0

    def __ge__(self, other):
        return self.compare(other) >= 0

    def __gt__(self, other):
        return self.compare(other) == 1


class ICFGWTOComp:
    def __init__(self, node: ICFGNode):
        self.node = node


    def getICFGNode(self) -> ICFGNode:
        return self.node

    @abstractmethod
    def accept(self, visitor):
        pass


class ICFGWTONode(ICFGWTOComp):
    def __init__(self, node: ICFGNode):
        self.node = node

    def accept(self, visitor):
        visitor.visitNode(self)


class ICFGWTOCycle(ICFGWTOComp):
    def __init__(self, head: ICFGWTONode, components: list[ICFGWTOComp]):
        self.head = head
        self.components = components

    def accept(self, visitor):
        visitor.visit(self)


class ICFGWTO:
    class WTOCycleDepthBuilder:
        def __init__(self, node_to_wto_cycle_depth):
            self.wto_cycle_depth = WTOCycleDepth()
            self.node_to_wto_cycle_depth = node_to_wto_cycle_depth

        def visit(self, cycle: ICFGWTOCycle):
            head = cycle.head.getICFGNode()
            previous_cycle_depth = self.wto_cycle_depth
            self.node_to_wto_cycle_depth[head] = self.wto_cycle_depth
            self.wto_cycle_depth = WTOCycleDepth()
            self.wto_cycle_depth.add(head)
            for component in cycle.components:
                component.accept(self)
            self.wto_cycle_depth = previous_cycle_depth

        def visitNode(self, node: ICFGWTONode):
            self.node_to_wto_cycle_depth[node.getICFGNode()] = self.wto_cycle_depth


    def __init__(self, graph: ICFG, entry: ICFGNode, scc=None):
        self.graph = graph
        self.entry = entry
        # Interprocedural WTO: `scc` is the set of FunObjVar *ids* in this
        # call-graph SCC.  Call edges into a callee in the same SCC are then
        # followed (becoming back-edges), so a (mutually) recursive function's
        # entry shows up as a WTO cycle head -- exactly like the C++ ICFGWTO.
        # If no SCC is given, the SCC is the entry's own function.
        if scc:
            self.scc_fun_ids = set(scc)
        else:
            self.scc_fun_ids = {entry.getFun().getId()}
        self.components: list[ICFGWTOComp] = []
        self.all_components : set[ICFGWTOComp] = set()
        self.head_ref_to_cycle: dict[ICFGNode, ICFGWTOCycle] = {}
        self.node_to_depth: dict[ICFGNode, int] = {}

        self._num = 0
        self._CDN: dict[ICFGNode, int] = {}
        self._stack: list[ICFGNode] = []

    def init(self):
        self.visit(self.entry, self.components)
        self._CDN.clear()
        self._stack.clear()
        self.build_node_to_depth()

    def component(self, node: ICFGNode) -> ICFGWTOCycle:
        partition = []
        for succ in self.get_successors(node):
            if self._CDN.get(succ, 0) == 0:
                self.visit(succ, partition)
        head = ICFGWTONode(node)
        ptr = ICFGWTOCycle(head, partition)
        self.head_ref_to_cycle[node] = ptr
        return ptr

    def visit(self, node: ICFGNode, components: list[ICFGWTOComp]):
        head = 0 # CycleDepthNumber head(0)
        min = 0 # CycleDepthNumber min(0)
        loop = False # bool loop
        self._stack.append(node)    # push(node)
        self._num += 1  # _num += CycleDepthNumber(1)
        head = self._num # head = _num
        self._CDN[node] = head # setCDN(node, head)
        for succ in self.get_successors(node): # forEachSuccessor(node, [&](const NodeT* succ)
            succ_dfn = self._CDN.get(succ, 0) # CycleDepthNumber succ_dfn = getCDN(succ)
            if succ_dfn == 0: # if (succ_dfn == CycleDepthNumber(0))
                min = self.visit(succ, components) # min = visit(succ, partition)
            else:
                min = succ_dfn # min = succ_dfn
            if min <= head: # if (min <= head)
                head = min # head = min
                loop = True # loop = true

        if head == self._CDN[node]: # if (head == getCDN(node))
            self._CDN[node] = 0x7fffffff # setCDN(node, UINT_MAX)
            element = self._stack.pop()
            if loop:
                while element != node:
                    self._CDN[element] = 0 # setCDN(element, 0)
                    element = self._stack.pop() # element = pop()
                components.insert(0, self.component(node)) # partition.push_front(component(node))
            else:
                components.insert(0, ICFGWTONode(node)) # partition.push_front(newNode(node))
        return head

    def get_successors(self, node: ICFGNode) -> list[ICFGNode]:
        # Interprocedural successor relation, mirroring C++ ICFGWTO::getSuccessors.
        successors = []
        if isinstance(node, pysvf.CallICFGNode):
            for e in node.getOutEdges():
                callee_entry = e.getDstNode()
                if callee_entry.getFun().getId() in self.scc_fun_ids:
                    # caller & callee in the same SCC -> follow the call edge
                    successors.append(callee_entry)
                else:
                    # different SCC -> shortcut to the local return node
                    successors.append(node.getRetICFGNode())
        else:
            for e in node.getOutEdges():
                succ = e.getDstNode()
                # Only stay within the SCC (intra edges, and return edges back
                # into an SCC function).
                if succ.getFun().getId() in self.scc_fun_ids:
                    successors.append(succ)
        return successors


    def build_node_to_depth(self):
        builder = self.WTOCycleDepthBuilder(self.node_to_depth)
        for component in self.components:
            component.accept(builder)


    def __str__(self):
        return f"ICFGWTO: {self.components}"


class AbstractExecution(ABC):
    def __init__(self, pag: pysvf.SVFIR):
        self.svfir = pag
        self.icfg = pag.getICFG()
        self.call_site_stack = []
        self.func_to_wto = {}
        self.pre_abs_trace = {}
        self.post_abs_trace = {}
        self.buf_overflow_helper = AEReporter(self.svfir)

        self.widen_delay = 3
        self.addressMask = 0x7f000000
        self.flippedAddressMask = (self.addressMask^0xffffffff)

    # =========================================================================
    # FUNCTION STUBS TO IMPLEMENT FOR ASSIGNMENT 3
    # =========================================================================

    @abstractmethod
    def handleGlobalNode(self):
        # TODO
        pass

    @abstractmethod
    def handleFunction(self, funEntry: pysvf.ICFGNode):
        # TODO
        pass

    @abstractmethod
    def handleICFGNode(self, node: pysvf.ICFGNode) -> bool:
        # TODO
        return False

    @abstractmethod
    def handleICFGCycle(self, cycle):
        # TODO
        pass

    @abstractmethod
    def handleCallSite(self, node: pysvf.CallICFGNode):
        # TODO
        pass

    @abstractmethod
    def bufOverflowDetection(self, node: pysvf.ICFGNode):
        # TODO
        pass

    @abstractmethod
    def nullptrDerefDetection(self, node: pysvf.ICFGNode):
        # TODO
        pass

    # =========================================================================
    # END ASSIGNMENT 3 FUNCTION STUBS
    # =========================================================================

    def getAEState(self, node: pysvf.ICFGNode) -> AEState:
        if node not in self.post_abs_trace:
            self.post_abs_trace[node] = AEState()
        return self.post_abs_trace[node]

    def _storePostState(self, node: pysvf.ICFGNode,
                        state: pysvf.AbstractState):
        self.post_abs_trace[node] = AEState(unwrap_state(state).clone())

    def analyse(self):
        self.initWto()
        self.handleGlobalNode()

        main_fun = self.svfir.getFunObjVar("main")
        assert main_fun, "Main function not found"
        entry = self.icfg.getFunEntryICFGNode(main_fun)
        entry_state = self.getAEState(
            self.icfg.getGlobalICFGNode()).clone()
        for i in range(main_fun.arg_size()):
            entry_state[main_fun.getArg(i).getId()] = IntervalValue.top()
        self.pre_abs_trace[entry] = AEState(entry_state.clone())
        self._storePostState(entry, entry_state)
        self.handleFunction(entry)
        self.ensureAllAssertsValidated()
        self.buf_overflow_helper.printReport()

    def reportBufOverflow(self, node, msg=None):
        self.buf_overflow_helper.reportBufOverflow(
            node, msg if msg is not None else f"buffer-overflow at {node}")

    def reportNullDeref(self, node, msg=None):
        self.buf_overflow_helper.reportBufOverflow(
            node, msg if msg is not None else f"nullptr-deref at {node}")

    def initWto(self):
        """
        Initialize the interprocedural WTO per call-graph SCC entry.
    
        Each (mutually) recursive function's entry node becomes a WTO cycle head
        because intra-SCC call edges are turned into back-edges.  The same
        widening/narrowing machinery used for loops then drives recursion to a
        fixpoint via handleICFGCycle; there is no separate "is recursive?" check.
        """
        callgraphScc = pysvf.getCallGraphSCC()
        self._callgraph_scc = callgraphScc
        callgraph = self.svfir.getCallGraph()
        self._callgraph = callgraph

        # SCC membership comes from pysvf: CallGraphSCC.subNodes(rep) returns
        # the call-graph node IDs in the SCC represented by 'rep'.  We only
        # need it to feed ICFGWTO so intra-SCC call edges become back-edges.
        cgid_to_fun = {}
        for node in callgraph.getNodes():
            cgid_to_fun[node.getId()] = node.getFunction()

        # Build one interprocedural WTO per call-graph-SCC entry function.  An
        # SCC entry is a member with a caller outside the SCC (or no caller).
        # Intra-SCC members are reached via the entry's interprocedural WTO.
        self.func_to_wto = {}
        for node in callgraph.getNodes():
            fun = node.getFunction()
            if fun.isDeclaration():
                continue
            cgid = node.getId()
            rep = callgraphScc.repNode(cgid)
            scc_cgids = set(callgraphScc.subNodes(rep))

            in_edges = list(node.getInEdges())
            is_entry = (len(in_edges) == 0)
            for e in in_edges:
                if e.getSrcID() not in scc_cgids:   # caller outside the SCC
                    is_entry = True
            if not is_entry:
                continue

            func_scc_ids = {cgid_to_fun[c].getId() for c in scc_cgids}
            wto = ICFGWTO(self.icfg, self.icfg.getFunEntryICFGNode(fun), func_scc_ids)
            wto.init()
            # Key by function id: pybind FunObjVar wrappers are not guaranteed to
            # hash consistently across calls, so don't use the object as a key.
            self.func_to_wto[fun.getId()] = wto

    def getVirtualMemAddress(self, idx: int) -> int:
        return self.addressMask + idx

    def inSameCallGraphSCC(self, fun1, fun2) -> bool:
        scc = getattr(self, "_callgraph_scc", None)
        cg = getattr(self, "_callgraph", None)
        if scc is None or cg is None:
            return False
        n1 = cg.getCallGraphNodeByFunObj(fun1)
        n2 = cg.getCallGraphNodeByFunObj(fun2)
        if n1 is None or n2 is None:
            return False
        return scc.repNode(n1.getId()) == scc.repNode(n2.getId())

    # Whitelist of external-call names covered by the assignment. Mirrors the
    # C++ side: exact match for assignment-specific and checkpoint stubs;
    # substring match for library APIs whose Clang lowering yields LLVM
    # intrinsics (e.g. `llvm.memcpy.*`).
    _EXT_EXACT_STUBS = frozenset({
        "mem_insert", "str_insert",
        "UNSAFE_BUFACCESS", "SAFE_BUFACCESS",
        "UNSAFE_PTRDEREF", "SAFE_PTRDEREF",
    })
    _EXT_API_SUBSTRINGS = (
        "memcpy", "memmove", "memset",
        "strcpy", "strncpy", "strcat", "strncat",
        "strlen", "wcslen",
    )

    def isExternalCallForAssignment(self, func) -> bool:
        name = func.getName()
        if name in self._EXT_EXACT_STUBS:
            return True
        return any(key in name for key in self._EXT_API_SUBSTRINGS)

    def handleStubFunction(self, callNode: pysvf.CallICFGNode):
        """
        Handle stub functions such as 'svf_assert' and 'OVERFLOW'.

        This function processes specific stub functions in the program's control flow graph (CFG) 
        to validate assertions or detect buffer overflows. It performs the following tasks:

        1. For 'svf_assert':
           - Adds the call node to the set of assertion points.
           - Checks the abstract state of the argument to determine if the assertion is valid.
           - If the assertion is invalid or unsatisfiable, raises an error.

        2. For 'OVERFLOW':
           - Adds the call node to the set of assertion points.
           - Checks if the right-hand side (RHS) value is an address.
           - Iterates through the addresses to calculate the access offset and compare it 
             with the object size to detect buffer overflows.
           - If a buffer overflow is detected, records the overflow node and prints a success message.
           - If no overflow is detected, raises an error.

        :param call_node: The call node representing the stub function in the CFG.
        :type call_node: pysvf.CallICFGNode
        """
        # Get the callee function associated with the call site
        if callNode.getCalledFunction().getName() == "svf_assert":
            self.buf_overflow_helper.noteAssertionPoint(callNode)
            # If the condition is false, the program is infeasible
            arg0 = callNode.getArgument(0).getId()
            abstract_state = self.post_abs_trace[callNode]

            # Check if the interval for the argument is infinite
            if abstract_state[arg0].getInterval().isTop():
                print(f"svf_assert Fail. {callNode}")
                assert False
            else:
                if (abstract_state[arg0].getInterval().equals(IntervalValue(1, 1)) or
                        abstract_state[arg0].getInterval().equals(IntervalValue(-1, -1))):
                    print(f"The assertion ({callNode}) is successfully verified!!")
                else:
                    print(f"The assertion ({callNode}) is unsatisfiable!!")
                    assert False
        elif callNode.getCalledFunction().getName() == "svf_assert_eq":
            self.buf_overflow_helper.noteAssertionPoint(callNode)
            arg0 = callNode.getArgument(0).getId()
            arg1 = callNode.getArgument(1).getId()
            abstract_state = self.post_abs_trace[callNode]
            if abstract_state[arg0].getInterval().equals(
                    abstract_state[arg1].getInterval()):
                print("The assertion is successfully verified!!")
            else:
                print(f"svf_assert_eq Fail. {callNode}")
                assert False

    def handleCheckpointStubs(self, callNode: pysvf.CallICFGNode):
        """SAFE_/UNSAFE_ checkpoints: ground-truth bug markers.

        Records the call site in ``assert_points`` so
        :py:meth:`ensureAllAssertsValidated` can verify the student's control
        flow reached it.  The harness reports a bug iff its independent
        ground-truth check (bypassing the student's predicates) sees one.
        """
        self.buf_overflow_helper.noteAssertionPoint(callNode)
        fun_name = callNode.getCalledFunction().getName()
        abstract_state = self.post_abs_trace[callNode]
        if fun_name in ("SAFE_BUFACCESS", "UNSAFE_BUFACCESS"):
            if callNode.arg_size() < 2:
                return
            length = abstract_state[callNode.getArgument(1).getId()].getInterval()
            if length.isBottom():
                length = IntervalValue(0)
            ptr = callNode.getArgument(0)
            if not self._harnessSafeAccess(abstract_state, ptr, length - IntervalValue(1)):
                self.buf_overflow_helper.reportBufOverflow(
                    callNode, f"buffer-overflow at {callNode}")
        elif fun_name in ("SAFE_PTRDEREF", "UNSAFE_PTRDEREF"):
            if callNode.arg_size() < 1:
                return
            ptr = callNode.getArgument(0)
            if not self._harnessSafeDeref(abstract_state, ptr):
                self.buf_overflow_helper.reportBufOverflow(
                    callNode, f"nullptr-deref at {callNode}")

    def _harnessSafeAccess(self, abstract_state, value, length: IntervalValue) -> bool:
        ptr_val = abstract_state[value.getId()]
        if not ptr_val.isAddr():
            return True
        for addr in ptr_val.getAddrs():
            if pysvf.AbstractState.isBlackHoleObjAddr(addr) or pysvf.AbstractState.isNullMem(addr):
                continue
            obj_id = abstract_state.getIDFromAddr(addr)
            base_obj = self.svfir.getBaseObject(obj_id)
            if base_obj is None or base_obj.isBlackHoleObj() or not base_obj.isConstantByteSize():
                continue
            size = base_obj.getByteSizeOfObj()
            gnode = self.svfir.getGNode(obj_id)
            base_offset = IntervalValue(gnode.getConstantFieldIdx()) if isinstance(gnode, pysvf.GepObjVar) else IntervalValue(0)
            offset = base_offset + length
            if int(offset.ub()) >= size:
                return False
        return True

    def _harnessSafeDeref(self, abstract_state, value) -> bool:
        if value is None or isinstance(value, pysvf.ConstNullPtrValVar):
            return False
        abs_val = abstract_state[value.getId()]
        if not abs_val.isAddr():
            return True
        for addr in abs_val.getAddrs():
            if pysvf.AbstractState.isBlackHoleObjAddr(addr):
                continue
            if pysvf.AbstractState.isNullMem(addr):
                return False
            if abstract_state.isFreedMem(addr):
                return False
        return True
    
    def ensureAllAssertsValidated(self):
        """Verify the student's control flow reached every ground-truth stub.

        Recognised stubs:
          * ``svf_assert`` / ``svf_assert_eq`` -- abstract-state assertions
          * ``UNSAFE_PTRDEREF`` / ``SAFE_PTRDEREF`` -- null-deref ground truth
          * ``UNSAFE_BUFACCESS`` / ``SAFE_BUFACCESS`` -- buffer-access ground truth

        A missed stub site means the student's control-flow logic skipped a
        place the grader cares about.  Additionally requires that the number
        of reported bugs is at least the number of ``UNSAFE_*`` stubs.
        """
        assert_stubs = {"svf_assert", "svf_assert_eq"}
        checkpoint_stubs = {"UNSAFE_PTRDEREF", "SAFE_PTRDEREF",
                            "UNSAFE_BUFACCESS", "SAFE_BUFACCESS"}
        unsafe_to_be_verified = 0
        for node in self.svfir.getICFG().getNodes():
            if not isinstance(node, pysvf.CallICFGNode):
                continue
            called_function = node.getCalledFunction()
            if not called_function:
                continue
            name = called_function.getName()
            if name not in assert_stubs and name not in checkpoint_stubs:
                continue
            if name.startswith("UNSAFE_"):
                unsafe_to_be_verified += 1
            if not self.buf_overflow_helper.isAssertionPoint(node):
                raise AssertionError(
                    f"The stub function callsite ({name}) was not reached by "
                    f"the student's control flow: {node}"
                )
        assert unsafe_to_be_verified <= len(self.buf_overflow_helper.node_to_bug_info), \
            "The number of UNSAFE_* stubs (ground truth) should <= the number of bugs reported"

    def updateStateOnCall(self, call: pysvf.CallPE):
        node = call.getICFGNode()
        abstract_state = self.post_abs_trace[node]
        joined = AbstractValue()
        for index in range(call.getOpVarNum()):
            call_node = call.getOpCallICFGNode(index)
            if call_node not in self.post_abs_trace:
                continue
            joined.join_with(
                self.post_abs_trace[call_node][call.getOpVarId(index)])
        abstract_state[call.getResId()] = joined
